### PR TITLE
[Dashboard] Use a dedicated RPC to check the GCS is alive

### LINKED ISSUE #16245: [Dashboard] Use a dedicated RPC to check if GCS is alive
Currently, dashboard use the `GetAllNodeInfo` RPC to check if GCS is alive which may false reporting the GCS is not alive. 

- Add a new RPC `CheckAlive` to the GCS.
- Handle the `CheckAlive` RPC in heartbeat thread.

For feature requests or questions, post on our Discussion page instead: https://discuss.ray.io/

### PR DESCRIPTION
**Pull request by fyrestone**
## Why are these changes needed?

- Dashboard use the GetAllNodeInfo RPC to check if GCS is alive which may false reporting the GCS is not alive.
- Check GCS is alive is a framework logic. It should be moved from node module to the dashboard head. 

So, this PR contains
- Add a new RPC CheckAlive to the GCS.
- Handle the CheckAlive RPC in heartbeat thread.
- Move the check GCS is alive logic from node module to dashboard head.
- Rename `LogicalViewHead` to `SnapshotHead` in snapshot_head.py
- Use `os._exit()` to avoid dashboard hangs at exit after the GCS is exit.

## Related issue number

Closes https://github.com/ray-project/ray/issues/16245
https://github.com/ray-project/ray/issues/16328

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
This is a nice change! @wuisawesome didn't we already add something similar? Can you follow up?

**Comment by wuisawesome** [comment]
Yeah would be great if we could make `ray health-check` use this mechanism too (it currently just directly pings redis).

File: src/ray/rpc/gcs_server/gcs_rpc_client.h [hunk 1]
```diff
@@ -193,6 +193,10 @@ class GcsRpcClient {
   VOID_GCS_RPC_CLIENT_METHOD(HeartbeatInfoGcsService, ReportHeartbeat,
                              heartbeat_info_grpc_client_, )
 
+  /// Check GCS is alive.
+  VOID_GCS_RPC_CLIENT_METHOD(HeartbeatInfoGcsService, CheckAlive,
```
Comment (wuisawesome) [inline src/ray/rpc/gcs_server/gcs_rpc_client.h:197]: Shouldn't `GcsNodeManager` be responsible for checking if a node is alive?

**Review by wuisawesome — CHANGES_REQUESTED** [review]
Can we update `ray health-check` too?

**Comment by rkooo567** [comment]
@wuisawesome one of side effects of that change is that the health check is probably not available from the default package (because the dashboard is an optional package)? Or are we using the dashboard to get the GCS health info right now as well?

**Comment by wuisawesome** [comment]
We shouldn't need to use the dashboard for the health check. i think we should be able to just directly query GCS w/ a python grpc client.

File: src/ray/rpc/gcs_server/gcs_rpc_client.h (hunk 1 above)
Comment (fyrestone) [reply src/ray/rpc/gcs_server/gcs_rpc_client.h:199]: This RPC is to check the head is alive.

Comment (fyrestone) [reply src/ray/rpc/gcs_server/gcs_rpc_client.h:199]: We have met the false reporting issue. So, here comes the improvement.

**Comment by fyrestone** [comment]
@rkooo567 @wuisawesome Is there any problem about this PR? It is pending review for a month.

**Comment by wuisawesome** [comment]
Can you add the `ray health-check` endpoint?

**Comment by rkooo567** [comment]
Let's merge it after Alex' comment is addressed

**Comment by fyrestone** [comment]
> Can you add the `ray health-check` endpoint?

Let me clarify this feature,
1. New `ray health-check` command.
2. It requests an RPC `CheckAlive` to the GCS, and checks the response. 

Did I miss something?

**Comment by wuisawesome** [comment]
Exactly. In particular, can we make sure it queries GCS directly and not through the dashboard? 

We should also switch the existing health checks that use redis, but that is out of scope for this PR.

**Comment by fyrestone** [comment]
> Exactly. In particular, can we make sure it queries GCS directly and not through the dashboard?
> 
> We should also switch the existing health checks that use redis, but that is out of scope for this PR.

Got it. I will update this PR for the `ray health-check`.

**Comment by fyrestone** [comment]
@wuisawesome @rkooo567 This PR is ready to be merged.

**Review by wuisawesome — APPROVED** [review]
Whoops sorry lost track of this one.

**Comment by wuisawesome** [comment]
Can you fix the conflict then merge?

File: src/ray/protobuf/gcs_service.proto [hunk 2]
```diff
@@ -189,6 +189,15 @@ message ReportHeartbeatReply {
   GcsStatus status = 1;
 }
 
+message CheckAliveRequest {
+  int32 seq = 1;
+}
+
+message CheckAliveReply {
+  GcsStatus status = 1;
+  int32 seq = 2;
```
Comment (ijrsvt) [inline src/ray/protobuf/gcs_service.proto:198]: Is `status` ever used?

File: dashboard/consts.py [hunk 3]
```diff
@@ -9,6 +11,13 @@
 ORGANIZE_DATA_INTERVAL_SECONDS = 2
 REDIS_KEY_DASHBOARD_RPC = "dashboard_rpc"
 REDIS_KEY_GCS_SERVER_ADDRESS = "GcsServerAddress"
+# GCS check alive
+GCS_CHECK_ALIVE_MAX_COUNT_OF_RPC_ERROR = env_integer(
+    "GCS_CHECK_ALIVE_MAX_COUNT_OF_RPC_ERROR", 10)
+GCS_CHECK_ALIVE_INTERVAL_SECONDS = env_integer(
+    "GCS_CHECK_ALIVE_INTERVAL_SECONDS", 5)
+GCS_RETRY_CONNECT_INTERVAL_SECONDS = env_integer(
+    "GCS_RETRY_CONNECT_INTERVAL_SECONDS", 2)
```
Comment (ijrsvt) [inline dashboard/consts.py:20]: Sorry, where is this used?

File: dashboard/tests/conftest.py [hunk 4]
```diff
@@ -52,3 +52,14 @@ def small_event_line_limit():
...
+
+@pytest.fixture
+def fast_gcs_failure_detection():
+    os.environ["GCS_CHECK_ALIVE_MAX_COUNT_OF_RPC_ERROR"] = "2"
+    os.environ["GCS_CHECK_ALIVE_INTERVAL_SECONDS"] = "1"
+    os.environ["GCS_RETRY_CONNECT_INTERVAL_SECONDS"] = "1"
+    yield
+    os.environ.pop("GCS_CHECK_ALIVE_MAX_COUNT_OF_RPC_ERROR", None)
+    os.environ.pop("GCS_CHECK_ALIVE_INTERVAL_SECONDS", None)
+    os.environ.pop("GCS_RETRY_CONNECT_INTERVAL_SECONDS", None)
```
Comment (ijrsvt) [inline dashboard/tests/conftest.py:65]: Can you make sure to restore the state properly here (set the value after the yield to the initial value).

File: dashboard/tests/test_dashboard.py [hunk 5]
```diff
@@ -641,5 +641,24 @@ def test_dashboard_port_conflict(ray_start_with_dashboard):
...
+    gcs_server_info = all_processes[ray_constants.PROCESS_TYPE_GCS_SERVER][0]
+    gcs_server_proc = psutil.Process(gcs_server_info.process.pid)
+
+    assert dashboard_proc.status() in [
+        psutil.STATUS_RUNNING, psutil.STATUS_SLEEPING, psutil.STATUS_DISK_SLEEP
+    ]
+
+    gcs_server_proc.kill()
+    gcs_server_proc.wait()
+    dashboard_proc.wait(10)
```
Comment (ijrsvt) [inline dashboard/tests/test_dashboard.py:660]: Can you add an assert that the dashboard exits with a non-zero status?

File: dashboard/tests/conftest.py (hunk 4 above)
Comment (fyrestone) [reply dashboard/tests/conftest.py:65]: Sure, save the old value and restore it after yield is better. I will fix it.

File: dashboard/tests/test_dashboard.py (hunk 5 above)
Comment (fyrestone) [reply dashboard/tests/test_dashboard.py:660]: OK. I will refine the test case.

File: src/ray/protobuf/gcs_service.proto (hunk 2 above)
Comment (fyrestone) [reply src/ray/protobuf/gcs_service.proto:198]: Yes, it is used here: https://github.com/ray-project/ray/blob/19026f337eab2716b3a9e520e748f7f5458884ff/dashboard/head.py#L60
And, the status field follows the GCS rpc reply structure.

File: dashboard/consts.py (hunk 3 above)
Comment (fyrestone) [reply dashboard/consts.py:20]: Sorry, I have renamed the `CONNECT_GCS_INTERVAL_SECONDS` to `GCS_RETRY_CONNECT_INTERVAL_SECONDS` to make it more clear. I will fix the code.

**Review by ijrsvt — COMMENTED** [review]
LGTM now! Thanks for fixing!

**Pull request merged**